### ISSUE TITLE
Difference on PPO Losses

### ISSUE THREAD

**Issue by roosephu** [ray-project/ray#2233]
Hello, I noticed a small difference between your PPO implementation and OpenAI Baselines'. 
The Line 63 of In the file `ray/python/ray/rllib/ppo/loss.py` uses `tf.minimum` to compute the loss for value function, while the Line 35 of `baselines/baselines/ppo2/ppo2.py` uses `tf.maximum`. I think `tf.maximum` is more reasonable, so is it a bug or a feature?

**Comment by richardliaw** [comment]
The `minimum` is used there to take a lower bound of the value function objective. The main idea is such that the value function does not change too rapidly.

It is used in the same spirit as the clipped surrogate objective, where the minimum between the unclipped and clipped objective is taken (in the baselines version, their surrogate objective is negative, so they take the maximum there).

**Comment by roosephu** [comment]
Perhaps I didn't make it clear. 

Let's look at baselines' code.  For value function, the loss is `mean(maximum(vf_losses1, vf_losses2))`, which is `mean(maximum(square(vf - R), square(clip(vf) - R)))`. 

For your implementation, the loss of value function is `self.vf_loss = tf.minimum(self.vf_loss1, self.vf_loss2)`, which is `mininum(square(vf - value_target), square(clip(vf) - value_target))`. Note the difference here: you're using `minimum` and Baselines is using `maximum`. 

Probably it's better to take an example. Say, the true value is 0, and our previous value function gives 100 and our current value function gives 50. Assuming the clip range is 10, we've achieved our goal and shouldn't optimize anymore. However, the loss in your code is `minimum((50 - 0)**2, (90 - 0)**2))`, so it will continue pushing 50 to 0. As you can see, using maximum can solve this issue: the maximum is taken on (90 - 0)**2, so there is no gradient on our value function. 

However, your PG loss is the same as Baselines'. 

In Baselines, the PG loss is `pg_losses = mean(maximum(pg_losses, pg_losses2))` where `pg_losses` is like `(-advantage) * ratio` (notice the minus sign here). So overall `pg_losses` equals to `-mean(minimum(advantage * ratio, advantage * clip(ratio)))`. Please also note that their final loss is `loss = pg_loss - entropy * ent_coef + vf_loss * vf_coef`. 

In your code, the PG loss is `self.surr = tf.minimum(self.surr1, self.surr2) = minimum(advantage * ratio, advantage * clip(ratio))`. Your final loss is `-self.surr + ...`, thus PG related term equals to `-minimum(advantage * ratio, advantage * clip(ratio))`, which is the same as Baselines'.

**Comment by richardliaw** [comment]
Thanks, your example is helpful. I understand that the PG loss is the same and only previously mentioned for contrast.

`maximum` seems to be the right thing to apply; we'll push a fix. Thanks!

**Comment by ericl** [comment]
Fixed in https://github.com/ray-project/ray/pull/2366

**Comment by ericl** [comment]
So oddly, this breaks the Pendulum tuned example. Reverting it to tf.minimum() allows Pendulum to learn again. I don't see why tf.minimum() is correct though.

cc @eugenevinitsky
